clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/EventDiffCalculatorTest.java

```diff
@@ -316,4 +332,84 @@ public void too_low_distributor_node_ratio_cluster_down_reason_emits_correspondi
+        EventFixture f = EventFixture.createForNodes(3)
+                .clusterStateBefore("distributor:3 storage:3")
+                .derivedClusterStateBefore("default", "distributor:3 storage:3")
+                .clusterStateAfter("distributor:3 storage:3")
+                .derivedClusterStateAfter("default", "distributor:3 storage:3 .1.s:m")
+                .derivedStorageNodeReasonAfter("default", 1, NodeStateReason.MAY_HAVE_MERGES_PENDING);
+
+        List<Event> events = f.computeEventDiff();
+        assertThat(events.size(), equalTo(2));
+        assertThat(events, hasItem(allOf(
+                eventForNode(storageNode(1)),
+                nodeEventForBucketSpace("default"),
+                nodeEventWithDescription("Altered node state in cluster state from 'U' to 'M'"))));
+        assertThat(events, hasItem(allOf(
+                eventForNode(storageNode(1)),
+                nodeEventForBucketSpace("default"),
+                nodeEventWithDescription("Node may have merges pending"))));
+    }
+
+    @Test
+    public void may_have_merges_pending_down_edge_event_emitted_if_derived_bucket_space_state_differs_from_baseline() {
+        EventFixture f = EventFixture.createForNodes(3)
+                .clusterStateBefore("distributor:3 storage:3")
+                .derivedClusterStateBefore("default", "distributor:3 storage:3 .1.s:m")
+                .derivedStorageNodeReasonBefore("default", 1, NodeStateReason.MAY_HAVE_MERGES_PENDING)
+                .clusterStateAfter("distributor:3 storage:3")
+                .derivedClusterStateAfter("default", "distributor:3 storage:3");
+
+        List<Event> events = f.computeEventDiff();
+        assertThat(events.size(), equalTo(2));
+        assertThat(events, hasItem(allOf(
+                eventForNode(storageNode(1)),
+                nodeEventForBucketSpace("default"),
+                nodeEventWithDescription("Altered node state in cluster state from 'M' to 'U'"))));
+        assertThat(events, hasItem(allOf(
+                eventForNode(storageNode(1)),
+                nodeEventForBucketSpace("default"),
+                nodeEventWithDescription("Node no longer have merges pending"))));
+    }
+
+    @Test
+    public void both_baseline_and_derived_bucket_space_state_events_are_emitted() {
+        EventFixture f = EventFixture.createForNodes(3)
+                .clusterStateBefore("distributor:3 storage:3")
+                .derivedClusterStateBefore("default", "distributor:3 storage:3")
+                .clusterStateAfter("distributor:3 storage:3 .0.s:m")
+                .derivedClusterStateAfter("default", "distributor:3 storage:3 .1.s:m");
+
+        List<Event> events = f.computeEventDiff();
+        assertThat(events.size(), equalTo(2));
+        assertThat(events, hasItem(allOf(
+                eventForNode(storageNode(0)),
+                nodeEventForNullBucketSpace(),
```

Comment:
Would `nodeEventForBaseline` be a more descriptive name here? Bucket space being null seems more of an implementation detail.